### ISSUE TITLE
Usage of serverSideApply as a replacement for createOrReplace

### ISSUE THREAD

**Issue by shawkins** [fabric8io/kubernetes-client#5215]
Relates to #5176

Doing a conversion of createOrReplace to serverSideApply with Keycloak makes me think that adding an unlock method, which we have talked about in the past, is a good idea.  That would make the full replacement:

resource.unlock().forceConflicts().serverSideApply()

A more nuanced issue is that kubernetes will sometimes update the managedFields time even if there are no other changes in the resource, causing a new revision that would not be created via just an update.  Logic that is sensitive to new revisions, such as operators, would need filtering to address this. 

@csviri - is this something that has already been seen with the operator sdk?

I can't find any information on how to avoid this, in fact it looks like a kubernetes bug based upon the documentation https://github.com/kubernetes/kubernetes/blob/6f116dc4d5ce996aac4a9ad48a16e72b059f9439/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go#L1253 

cc @manusa

**Comment by csviri** [comment]
> A more nuanced issue is that kubernetes will sometimes update the managedFields time even if there are no other changes in the resource, causing a new revision that would not be created via just an update. Logic that is sensitive to new revisions, such as operators, would need filtering to address this.

Currently I'm working on a feature to handle the `dependent resource` updates with SSA, and just using the patch api. But also on an approach to make sure that we only patch the resource if did not change. 
https://github.com/java-operator-sdk/java-operator-sdk/pull/1928

So thepretically, this won't be a problem, the matching is based on `managedFields`. So we do an patch only if the managed fields by the controller are changed vs the desired state. So we don't do unnecessary patch, so the time field is not an issue.

 In SSA ideally the only the desired state is sent, without resource version indeed. But when the desired state is constructed the resourceVersion is not filled anyways.
 So from this point I don't see a huge need for something like `unlock()`. But like the `forceConflicts` and the `serverSideApply` part. That makes this nicely explicit.

**Comment by shawkins** [comment]
> So theoretically, this won't be a problem (the matching is based on managedFields). So we do an patch only if the managed fields by the controller are changed vs the desired state. So we don't do unnecessary patch, so the time field is not an issue.

That's good to know you're already filtering for unnecessary patches.  

For non-java operator sdk usage I'll probably open an upstream issue to see if that is expected behavior and add a note somewhere with whatever comes out of that.

> So with SSA it is usually not the case that you make modifications on a read resources, rather send a resource with your managed fields filled only.

There's nothing prohibiting you from starting with the existing version - that was quite common in paths that use createOrReplace.  What I'm looking for is the simplest replacement that will work in all circumstances to make the migration as straight-forward as possible.

**Comment by csviri** [comment]
> There's nothing prohibiting you from starting with the existing version - that was quite common in paths that use createOrReplace. What I'm looking for is the simplest replacement that will work in all circumstances to make the migration as straight-forward as possible.

the problem what I see is that, you might send fields which were added by a different controller, when working with resources that are read from the server, and in that case would take over or become a manager of a field that you did not intended to manage. 

But of course this is just in case there are multiple controllers, or when dynamic admission controllers hooks adds fields etc.

**Comment by shawkins** [comment]
> the problem what I see is that, you might send fields which were added by a different controller, when working with resources that are read from the server, and in that case would take over or become a manager of a field that you did not intended to manage.

That is true.  I'm trying too hard to ease the migration pain, but it's not really that simple.  Here's the range of options:

1.a. Change our serverSideApply function to error, rather than strip, managedFields to enforce that it should not be used on resource modified directly from the server.  No unlock function would be offered - at least specifically for the purposes of server side apply.  We'll provide documentation on when it's appropriate to serverSideApply vs createOrReplace.
1.b. And/or offer an unlock function along with documentation for serverSideApply of why that is a bad idea in circumstances where you may end up taking ownership of fields you did not originally modify.  Such a function is applicable to update (the replacement for replace would be resource.unlock().update()) and other patch operations.
1.c. And/or add a more straight-forward createOr function - we've mentioned a couple of times about having a createOrEdit/Patch with greatly simplified mechanics and of course explicit resourceVersion locking behavior.  This will give us a more straight-forward replacement for createOrReplace, but is one off that I'm still hesitant to provide.

**Comment by shawkins** [comment]
Captured the upstream with additional revisions as https://github.com/kubernetes/kubernetes/issues/118519

**Comment by shawkins** [comment]
@manusa @csviri related to https://github.com/kubernetes/kubernetes/issues/118519#issuecomment-1589628070 and #5214 we should consider serializing empty strings as null to match the go handling.

**Comment by shawkins** [comment]
@csviri did hit another nuanced server side apply issue.  A service managed by the operator was manually modified to include a port name.  As a json patch the server does not complain and allows that to happen.  The server side apply logic though is set to just assert the port number - when it does the server instead complains with a 422 "cannot set both port name & port number".
